### PR TITLE
Add xx30-maximized and xx30-hotp-maximized boards (11.5mb flashable BIOS regions, reproducible me.bin and generated gbe.bin and totally externally and internally flashable roms)

### LINKED ISSUE #307: Provide IFD, Gbe and ME blobs
The 12 MB x230 ROM image isn't flashable since it doesn't include the IFD and ME sections.  This isn't normally a large problem for updates since `flashrom-x230` won't touch the ME region, but it does lead to problems for initial install.

----
tlaurion hijacking top post here to give actual status in a glimpse.

- @osresearch accepted that heads distribute extracted IFD from an original Lenovo's rom, lower down in this issue [here](https://github.com/osresearch/heads/issues/307#issuecomment-552961256). The goal of this is to not generate the IFD ourselves and just unlock the one being present in the repository, on which BIOS region has been extended resulting of freed ME region, where the ME region is reduced to [the minimal size possible](https://github.com/osresearch/heads/pull/703/files#diff-0493cf37597ff5795588f8343ab268b7R3-R4). This results in a BIOS region of 11.5mb instead of 7mb, mitigating #590 issue until help joins forces for boards with limited SPI flash space without the need of soldering...
- [3 collaborators decided to support the adventure](https://github.com/osresearch/heads/issues/307#issuecomment-613514816) of [not waiting for legal approval](https://github.com/osresearch/heads/issues/307#issuecomment-615403721), nor expecting [Intel reply](https://github.com/osresearch/heads/issues/307#issuecomment-583887020), and going forward and include the ME cleaned binary.  Some efforts are still needed so the blob is downloaded directly from Lenovo, on which innoextract needs to be applied to remove the blob itself from local blob directory and do the required magic in a script having downloaded the blob from Lenovo website directly. Meanwhile the cleaned blob is present in tree with manual reproducibility instructions in xx30's blobs's directory README. [Read the xx30 blobs README here for more details](https://github.com/osresearch/heads/pull/703/files#diff-0493cf37597ff5795588f8343ab268b7). Please show interest for being paid fixing this.
- [Along the adventure](https://github.com/osresearch/heads/issues/307#issuecomment-626882321), it was discovered and [documented upstream](https://github.com/osresearch/heads/issues/700#issuecomment-627208266) that the GBE region needs to be present to have a functional ethernet card. This GBE blob [could be totally generated](https://github.com/osresearch/heads/issues/307#issuecomment-646897081). But would need some low level people involved. I'm not good at [reading reference manual and implementing what is required](https://github.com/osresearch/heads/issues/700#issuecomment-647003701). Please show interest for being paid fixing this.

Mitigations from dropping those binaries would be to:
- [ ] [Claim right to repair laws](https://gizmodo.com/right-to-repair-is-less-complicated-and-more-important-1834672055)? 
- [ ] Claim [Fair use laws](https://djtechtools.com/2017/06/05/basics-dj-copyright-laws/)?
- [ ] Generate those blobs from source code (see below)

So the result of the ongoing #703 PR is a PoC of that, which requires Heads to fake MAC address on boot to not have a MAC collision on the same LAN network segment if many Heads xx30 empowered machines are present.

Todo:
- [ ] [Generate GBE blob ](https://github.com/osresearch/heads/issues/700#issuecomment-647003701) to remove the GBE blob from tree. Seperate issue: #796 
- [ ] [Extract ME from Lenovo website (innoextract) and run me_cleaner on top of it](https://github.com/osresearch/heads/pull/703/files#diff-0493cf37597ff5795588f8343ab268b7) to put resulting blob locally to remove it from tree. Seperate issue: #797

### PR DESCRIPTION
**Pull request by tlaurion**
**_REQUIRES_**:
- External flashing with individual `maximized-top.rom` on top SPI chip and `maximized-bottom.rom` counterpart on bottom SPI flash chip.
**_or_**
- Internal flashing over initially unlocked ifd (did you run `ifdtool -u` on your bottom.rom stock backup prior of reflashing it?).
If so, internal flashing is possible through a manual `flashrom --force --noverify-all -p internal -w maximized.rom` (12mb rom) from command line to switch from xx30 to xx30-maximized.

Adds:
- blobs/xx30/me.bin (98k, ROMP and BUP only modules, cannot be removed), shared among all xx30 boards (Helix (Type 3xxx), T430, T430i, T430s, T430si, T431s, T530, T530i, W530, X1 Carbon (Type 34xx), X1 Helix (Type 3xxx), X1 Helix (Type 3xxx) 3G, X230, X230i, X230 Tablet, X230i Tablet, X230s
- blobs/xx30/ifd.bin (~~Might be generalizeable to ifd.bin: haven't tested on above boards~~ _EDIT: Reported working_)
- adds coreboot config which [specifies maximal CBFS space usable inside of Heads (11.5Mb instead of 7Mb)](https://github.com/osresearch/heads/pull/703/files#diff-935e48e65275b2fcaaeedb57a41098b9R8)
- ~~GitlabCI~~ (**EDIT: changed their caching system. Unusable for our use case now**) and CirecleCI modifications, ~~changing the default board compiled to be x230-external-flash (instead of x230)~~ EDIT: no replace. 
- CI downloads [build dependencies](https://github.com/osresearch/heads-wiki/blob/master/Building.md), enforces cache ~~([GitlabCI](https://github.com/osresearch/heads/pull/703/files#diff-96edf7a6f008de9e928d04e1ae5e12a5R15)~~, CircleCI ([1](https://github.com/osresearch/heads/pull/703/files#diff-1d37e48f9ceff6d8030570cd36286a61R15))([2](https://github.com/osresearch/heads/pull/703/files#diff-1d37e48f9ceff6d8030570cd36286a61R15))) and quick fixes ([cleans make build](https://github.com/osresearch/heads/pull/703/files#diff-1d37e48f9ceff6d8030570cd36286a61R51) and[ board local builds](https://github.com/osresearch/heads/pull/703/files#diff-1d37e48f9ceff6d8030570cd36286a61R51) ~~that are not properly declared in Makefile... To fix later, just want it to work right now)~~

CI:
- Add build logs as artifact to more easily compare non reproducible builds.
- Add initrd.cpio.xz to facilitate build reproducibility troubleshooting between CIs builds.

Permits:
- deprecation of x230-flash board currently required for initial internal flashing of x230. 
- Internal flashing of the whole 12Mb flash image and not only the BIOS region. Requires fully valid ROM (no faked IFD as currently used for xx30 boards) for subsequent ROM updates, since[ board flashrom option does not specify BIOS region anymore](https://github.com/osresearch/heads/pull/703/files#diff-0d85bd26ac1f30e0b0a79b266deb1be1R39) for subsequent flashrom upgrades.
- [Maximal space](https://github.com/osresearch/heads/pull/703/files#diff-935e48e65275b2fcaaeedb57a41098b9R8) for other tools integration.

Tested on x230 and x230t from latest commit CI build's artifacts. (Click on green checkmark to go on CircleCI, go in the artifacts section and download ROMs (both top and bootom roms to externally flash, or, full 12mb rom to flash internally, see REQUIREMENTS)

Tagging all x230 owners as per #691 : @osresearch @merge @jan23 @flawedworld @MrChromebox @shamen123 @eganonoa
~~T430 owners: want to adapt and test if x230-ifd.bin can be used on that board too?~~ EDIT: reported to work.

Would close #307.

Edit: put links everywhere to facilitate understanding implied changes.

### CODE REVIEW

**Comment by tlaurion** [comment]
I could rename x230-external-flash board to x230-neutered_me if your prefer.
@osresearch : Following #307 discussions, it is impossible to download ifd, bugt possible to download ME. 

If you have any advice into putting innoextract usage into a module makefile, I could remove ME from the xx30 directory and only have x230-ifd.bin (that might generalize into ifd after other xx30 boards are tested.)

Waiting for your input guys!

**Comment by tlaurion** [comment]
@osresearch @merge @jan23 @flawedworld @MrChromebox @shamen123 @eganonoa
Please test and comment.

**Comment by tlaurion** [comment]
Will merge myself in the next coming days. This project stands still and this si really, really sad.

**Comment by tlaurion** [comment]
@MrChromebox : thanks for the review!

> why a separate linux config for this board? it's exactly the same as linux-x230.config -- why not just use that? 

Agreed. More modules inclusion in the main linux config to come. Actually, I would go forward and also upgrade the kernel to at least the latest LTS kernel of that branch, or even better, move kernel support to 5.x branch. Will do in seperate PR. The reasoning on this is that the x230 board has actually changed in time. For example, the MMC controller passed from Ricoh to Realtek. Unfortunately, Realtek MMC kernel support is not in the current kernel nor config. 

In the meantime I will point to the linux-x230.config

> I realize you just duped/edited the coreboot-x230 config, but there's a lot in there that doesn't make sense / isn't needed:
> -- why not include the config file in CBFS?

Added
> -- why not include timestamps in cbmem?

Added
> -- USE_BLOBS is useless since it only affects coreboot checking out the blobs repo, which we do manually

Added
> -- deselecting POST_IO and POST_DEVICE is better done via CONFIG_NO_POST=y

Removed
> -- makes no sense to select DRIVERS_UART_8250IO then disable CONSOLE_SERIAL

Removed 
> -- DRIVERS_PS2_KEYBOARD isn't needed for linux payload

Removed
> -- UART_PCI_ADDR is default and doesn't need to be set

Removed
> -- PXE is deselected by default, no need to explicitly deselect

Removed
> -- no need for DEBUG_SMM_RELOCATION

Removed
> the above changes drop the config from 29 lines to 19

Same result
> for the ROM split, any particular reason to use a 64kb block size, vs say 4M (assuming you wanted to use the same for both)?
Just reused an old recipe of Trammel that just worked (in tree) that didn't made any sense back then because of faked IFD and absence of ME, creating invalid 4Mb and 8Mb images. It doesn't change anything really, unless you have a concise answer to change it.

**Comment by tlaurion** [comment]
- Tested CI build produced above on x230t flashing both bottom.rom and top.rom: Works.
- Tested flash.sh to flash whole 12mb coreboot.rom internally: Works.

**Comment by tlaurion** [comment]
Received FSFE Legal advice today. [Read here](https://github.com/osresearch/heads/issues/307#issuecomment-615403721)

**Comment by tlaurion** [comment]
@MrChromebox : following the legal advice above. Is there any trick i've missed that would permit to use innoextract or cabextract against the ME executable to automatize this without having to host the blob? @osresearch was ok fdor github to distribute IFD but not so confortable distributing ME.

I have no clue how to accomplis this.

**Comment by MrChromebox** [comment]
@tlaurion you had ME extraction working from FL2 via innoextract per your comments in #307 [here](https://github.com/osresearch/heads/issues/307#issuecomment-599279810), is that not sufficient?

**Comment by tlaurion** [comment]
Process is documented in readme to reproduce me.bin, not automatized. As a result the me.bin is in the PR, not code that would produce that me.bin without hosting it.

On April 19, 2020 6:06:14 PM UTC, MrChromebox <notifications@github.com> wrote:
>@tlaurion you had ME extraction working from FL2 via innoextract per your comments in #307 [here](https://github.com/osresearch/heads/issues/307#issuecomment-599279810), is that not sufficient? 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/osresearch/heads/pull/703#issuecomment-616197549
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by tlaurion** [comment]
> @tlaurion you had ME extraction working from FL2 via innoextract per your comments in #307 [here](https://github.com/osresearch/heads/issues/307#issuecomment-599279810), is that not sufficient?

@mrchromebox: still trying to find a way. New attempt is to make a xx30 submodule depending on me_cleaner and have innoextract downloaded on host system (CIs). Trying it make it work.

**Comment by tlaurion** [comment]
Seems like [without gbe blob, no ethernet](https://github.com/osresearch/heads/issues/700#issuecomment-620004416).

@hughies? Could we extract that from LVFS archive from a download script without hosting the blobs ourselves anywhere?

**Comment by MrChromebox** [comment]
since this went in after the flashrom 1.2 bump, can update CONFIG_FLASHROM_OPTIONS to drop the unnecessary params now

**Comment by tlaurion** [comment]
@MrChromebox : Will change FLASHROM_OPTIONS, thanks for the heads up. But with gbe needing to be inserted in coreboot, there will still be work needed and live extraction and neutering+deactivation of ME and unlocking of ifd... There is still work needed to go forward witht htis PR:
1- Can we distributed Gbe ... 
2- Can it be extracted from online source (not horsted on Github) from CI, and be reinjected in ROM so that #700 is not a problem...
3- Can we fake the mac and inject it or wipe the mac from CI, [wipe gbe while still having a functional ehternet discovered by e1000e](https://github.com/osresearch/heads/issues/700#issuecomment-620410981)

**Comment by MrChromebox** [comment]
it seems like we're jumping thru a lot of hoops to try and build a reproducible full flash image, rather than a reproducible coreboot/Heads one. Any reason we don't simply focus on the BIOS region only, and let the IFD/ME/GBE be the user's (or a downstream) problem?

**Comment by tlaurion** [comment]
@MrChromebox The problem i'm attacking here is to be able to add more stuff in the BIOS region, which is actually limited to 7Mb and near saturation. 
- #562 and consequently https://github.com/osresearch/heads/pull/709 (see vboot-*-me boards) cannot have VBOOT added as a result as of now, which leads to #590 which will at best following readings I have done, without LTO in but whith "hammering" option, could lead to an expected 11% of additional space currently occupied by the kernel.
- If we dream a bit, and open the door to #689 or #202 support, or want to open the door to remotely/locally verify LVM or any other task linked to integrity/security audit (#231) this won't be possible, which is why u-root and other projects are becoming interesting as alternatives to Heads because of code being ran at run time (text files highly compressible) while Heads rely mainly on cryptsetup (u-root inclusion: not interested), gpg(u-root: not interested and point to other projects which would then depend on same native libraries for usb token support anyway).

My goal here is, once again, to give a chance to Heads to do what it actually does already well and stop relying on patches (coreboot vboot) to do its job, while maximizing available space for other tools integration, while permitting CIs to build those reproducible images it is supposed to produce out of the box.

Continuing the BIOS region path is not only limiting, but doesn't permit to neither internally validate signed OEM rom image completely, nor verify ME partitions against external source, nor would permit FWUPD to completely reflash/upgrade the whole 12mb flash, which would need to be consistent.

The truth is that more CBFS space is needed. [Proof here, now.](https://github.com/osresearch/heads/commit/4868694a6b072b7fd138d8bb476cbe7f774c4ea9). xx30 and xx20(even more) are stuck at coreboot 4.8.1 if we don't do something about it.

**Comment by tlaurion** [comment]
> it seems like we're jumping thru a lot of hoops to try and build a reproducible full flash image, rather than a reproducible coreboot/Heads one. Any reason we don't simply focus on the BIOS region only, and let the IFD/ME/GBE be the user's (or a downstream) problem?

The "reproducible" image per se will be alive when:
- [ ] https://github.com/osresearch/heads/issues/640 is fixed
- [ ] https://github.com/osresearch/heads/issues/571#issuecomment-596243303 is fixed
- [ ] https://github.com/osresearch/heads/pull/715 is merged 
- [ ] https://github.com/osresearch/heads/issues/571 is addressed as a whole, building the same board across different docker images and confirming outcome sha256sum somehow.

This is not attempting to address reproducibility per se, but distributivity of a valid 12mb image to be wholly flashable/verifyable internally (through heads) and externally (boot media with io=relaxed) while yes, repdoducible through CIs per above points.

 https://github.com/osresearch/heads/pull/715

**Comment by flawedworld** [comment]
Just glanced at this and if we distribute gbe.bin, the first thoughts to my mind would be setting the mac address in the distributed blob to be 00:00:00:00:00:00 to achieve reproducible builds. The user can then set the mac address within the OS. Alternatively using a hex editing script / guide we can show the user how to edit the mac address post / pre compilation. If this is irrelevant then sorry for being a bother.

**Comment by tlaurion** [comment]
@flawedworld got successful?

**Comment by tlaurion** [comment]
> > it seems like we're jumping thru a lot of hoops to try and build a reproducible full flash image, rather than a reproducible coreboot/Heads one. Any reason we don't simply focus on the BIOS region only, and let the IFD/ME/GBE be the user's (or a downstream) problem?
> 
> @MrChromebox The "reproducible" image per se will be accomplished when:
> 
>     * [x]  [librem-hotp-verification is made reproducible](https://github.com/osresearch/heads/issues/640)
> 
>     * [ ]  [kernel is made reproducible across diffferent docker images again](https://github.com/osresearch/heads/issues/571#issuecomment-596243303)
> 
>     * [x]  [CircleCI change in configuration is merged to permit qemu-coreboot investigations](https://github.com/osresearch/heads/pull/715)
> 
>     * [x]  [When this qemu-coreboot board will be merged](https://github.com/osresearch/heads/pull/707), duplicated for other uses-cases, and [bonified](https://github.com/osresearch/heads/issues/516#issuecomment-618194043) (to support virtual TPM, communicate with LibremKey, and streamline tests without having to flash boards)
> 
>     * [ ]  [Heads will have a configurable CI permitting to build for different docker images and boards, confirming outcome sha256sum upstream somehow and distributing built artifacts as releases for tags](https://github.com/osresearch/heads/issues/571).
> 
> 
> This PR is not attempting to address reproducibility per se, but distributivity of a valid 12mb x230 image to be wholly flashable/verifyable internally (through heads to deploy FWUPD updates) and externally (boot media with io=relaxed) while yes, reproducible through CIs per above points is the continuity on that path.

Updated https://github.com/osresearch/heads/pull/703#issuecomment-623697903 comment with updates. Kernel still needs to be made reproducible, CI configurable and coreboot-qemu coreboot-qemu-gui board bonified to be useful (USB security dongle passed through usb passthrough to qemu, virtual tpm enabled, external storage image linked so Heads dont complain about /boot not being found, etc)

**Comment by tlaurion** [comment]
https://github.com/osresearch/heads/issues/700#issuecomment-646805590

**Comment by swiftgeek** [comment]
You should use bincfg from coreboot repo, inch9gen is deprecated in favor of it.

**Comment by tlaurion** [comment]
Update with discussion with switfgeek.
ich9gen -> bincfg

There are two .set files in coreboot 4.8.1:
```
user@x230-master:~/heads/build/coreboot-4.8.1/util/bincfg$ ls *.set
gbe-ich9m.set  ifd-x200.set

```

**Comment by Thrilleratplay** [comment]
@tlaurion  While I am thinking about it, I am curious if the generated GBE can be made smaller by removing the trailing padding and resized in IFD.  This may provide a little extra space.  I need to focus on the wiki for now but will try to test that idea later.

**Comment by tlaurion** [comment]
Note: the lenovo ME exe was saved to archive.org and downloadable here https://web.archive.org/web/20200928141415if_/https://download.lenovo.com/pccbbs/mobiles/g1rg24ww.exe

**Comment by tlaurion** [comment]
> @tlaurion While I am thinking about it, I am curious if the generated GBE can be made smaller by removing the trailing padding and resized in IFD. This may provide a little extra space. I need to focus on the wiki for now but will try to test that idea later.

@Thrilleratplay GBE binary is 2k, I think this is not a space limitor at all as compared to #590

**Comment by tlaurion** [comment]
We have a successful PoC!!!!

Download/Extraction in CI takes one second! 

[Full build of all roms](https://app.circleci.com/pipelines/github/tlaurion/heads/598/workflows/202ad9c3-7e06-45d2-ac7e-cd3292ea596b/jobs/646/steps)
[Extraction neutering/measuring job](https://app.circleci.com/pipelines/github/tlaurion/heads/598/workflows/202ad9c3-7e06-45d2-ac7e-cd3292ea596b/jobs/646/parallel-runs/0/steps/0-147)
[Resulting x230-external-flash 12 mb rom](https://646-103208611-gh.circle-artifacts.com/0/build/x230-external-flash/heads-x230-external-flash-v0.2.0-968-g2e18ea9.rom)
Resulting [x230-external-flash-bottom.rom](https://646-103208611-gh.circle-artifacts.com/0/build/x230-external-flash/x230-external-flash-bottom.rom) 
Resulting [x230-external-flash-top.rom](https://646-103208611-gh.circle-artifacts.com/0/build/x230-external-flash/x230-external-flash-top.rom) 

Relative [commit](https://github.com/tlaurion/heads/commit/1280231b58ccf674751891f924ea9294ea5183bd)

**Comment by tlaurion** [comment]
per https://github.com/osresearch/heads/issues/692 t430 users: @flawedworld @Thrilleratplay @alexmaloteaux @nickfrostatx @lsafd @bwachter @Trapacid

last commit adds t430-external-flash board for testing (I do not own the board).
Resulting in [bottom.rom](https://654-103208611-gh.circle-artifacts.com/0/build/t430-external-flash/t430-external-flash-bottom.rom) and [top.rom](https://654-103208611-gh.circle-artifacts.com/0/build/t430-external-flash/t430-external-flash-top.rom) to flash internally, where [heads-t430-external-flash-v0.2.0-977-g1db69ff.rom](https://654-103208611-gh.circle-artifacts.com/0/build/t430-external-flash/heads-t430-external-flash-v0.2.0-977-g1db69ff.rom) can be internally upgradeable (full 12mb reflash, reflashing me.bin, gbe.bin and BIOS region.)

**_WARNING_**: 
- this changes your ethernet MAC to `DE:AD:C0:FF:EE`  persistently (OS needs to randomize it, not yet done under Heads) 
- This ROM needs to be flashed externally at first
- This rom [then reflashes your whole ROM afterward internally](https://github.com/osresearch/heads/pull/703/files#diff-d7e056deb459d8cce5d5988269fea48446083e3b542f0e4e7a25ba254db4fe15R42) (as opposed to [only flashing your BIOS region](https://github.com/osresearch/heads/blob/master/boards/t430/t430.config#L50))

File: .circleci/config.yml [hunk 1]
```diff
@@ -65,11 +65,11 @@ jobs:
             rm -rf build/librem_l1um/* build/log/* && make CPUS=4 V=1 BOARD=librem_l1um || touch /tmp/failed_build
           no_output_timeout: 3h
       - run:
-          name: Output build failing logs 
+          name: Ouput build failing logs
```
Comment (Thrilleratplay) [inline .circleci/config.yml:68]: @tlaurion  Search "Ouput" replace with "Output" in `.circleci.yaml`.  There are multiple instances.

**Comment by tlaurion** [comment]
@Thrilleratplay build attempt with x230 and t430 boards still present. Maybe I should deprecate old x230 boards altogether?

@merge @jan23 @flawedworld @MrChromebox @shamen123 @eganonoa @bwachter @Thrilleratplay @SebastianMcMillan (from https://github.com/osresearch/heads/issues/692) ?

Warning: external initial flashing **_MAY BE REQUIRED_**, since we do not know current status of ME and re-localization (ME just deactivated, or neutered + relocated), not knowing if ifd being unlocked or not. 
Basically, if `ifdtool -u` was called on backup initially prior of reflashing the 8mb bottom SPI chip, **you could adventurously** (IF YOU HAVE EXTERNAL FLASHER to flash the two ROMs individually to fix brick) flash internally the whole ROM with:
`flashrom --force --noverify-all -p internal -w heads-x230-hotp-maximized-rom_tag-git_commit.rom` that will be produced on CircleCI. Note that you have to choose between hotp-maximized or maximized versions, depending if you own a Nitrokey Pro/Nitrokey Storage/Librem Key that you intend to use for remote attestation (aside of its smartcard features).

**Comment by tlaurion** [comment]
For some reason, cache has expired so builds will take longer then expected under last commit linked CI, [happening under yellow dot now on last commit](https://app.circleci.com/pipelines/github/tlaurion/heads/621/workflows/2b86f18e-0ad4-4202-8872-94aa798301f3/jobs/669).

[When done, ROMs, hashes files, build logs and cpioz will be be found under the artifact section on CircleCI](https://app.circleci.com/pipelines/github/tlaurion/heads/621/workflows/2b86f18e-0ad4-4202-8872-94aa798301f3/jobs/669/artifacts)

**Comment by bwachter** [comment]
Is it possible to change the MAC before flashing with a tool like the old ich9gen, or is a local build required for that?

**Comment by Thrilleratplay** [comment]
@tlaurion While we have talked about the path to migrate Lenovo models  `xx20` and `xx30` to larger Heads BIOS partitions, there seems to be a confusion on what is going on.  Given the many moving parts, maybe create a Github project to track the progression of issues/PRs/Questions?

@bwachter As of coreboot 4.13, a `bincfg` similar to ich9gen [has been added to allow generation of the GBE](https://github.com/coreboot/coreboot/blob/master/util/bincfg/Makefile#L23) for `xx20` and `xx30` boards with the ability to change the MAC address.  This PR includes a patch which back ports it to coreboot 4.8. Although, this raises the concern that not everyone may want to use the default `00:DE:AD:C0:FF:EE`.  As it stands right now, the would need either need to build locally with the desired MAC created in the blobs directory or flash the user created GBE partition separately after Heads has been updated.  Perhaps this can be simplified for the user by the `flash.sh` generating a patch of the MAC and checksum from the current GBE to be applied to the written ROM.

**Comment by tlaurion** [comment]
As of right now @bwachter : you can change the mac [here](https://github.com/osresearch/heads/pull/703/files#diff-02ce7c881356c0d23b8c74928036719f645462fa735101f1e7e62d763ebc732dR61-R66).
And rerun the detailed instructions found [here](https://github.com/osresearch/heads/pull/703/files#diff-2713f9bba9c6bd279025474d76958633751480c8655f7c2ac77471498e6f7ea4), specifically the instructions to [manually generate gbe.bin](https://github.com/osresearch/heads/pull/703/files#diff-2713f9bba9c6bd279025474d76958633751480c8655f7c2ac77471498e6f7ea4R46-R57) or the [instructions to extract from backup your gbe.bin](https://github.com/osresearch/heads/pull/703/files#diff-2713f9bba9c6bd279025474d76958633751480c8655f7c2ac77471498e6f7ea4R15-R16)

**Comment by tlaurion** [comment]
> @tlaurion While we have talked about the path to migrate Lenovo models `xx20` and `xx30` to larger Heads BIOS partitions, there seems to be a confusion on what is going on. Given the many moving parts, maybe create a Github project to track the progression of issues/PRs/Questions?

Those were specified [here](https://github.com/osresearch/heads/pull/703#issuecomment-735883422), mainly:
"Basically, if ifdtool -u was called on backup initially prior of reflashing the 8mb bottom SPI chip, you could adventurously (IF YOU HAVE EXTERNAL FLASHER to flash the two ROMs individually to fix brick) flash internally the whole ROM with:
flashrom --force --noverify-all -p internal -w heads-x230-hotp-maximized-rom_tag-git_commit.rom that will be produced on CircleCI. Note that you have to choose between hotp-maximized or maximized versions, depending if you own a Nitrokey Pro/Nitrokey Storage/Librem Key that you intend to use for remote attestation (aside of its smartcard features)."

As for deprecating xx20 and xx30 boards, @flammit showed opposition.
There is also opposition in unlocking ifd, considering ME is not measured right now. 

Considering BIOS region is measured, sealed, and validated just before boot with Disk Unlock Key passphrase, i would go forward and even implement a flashrom verification of ME partition prior of booting into another PR. Otherwise, no other solution were proposed, and xx20 and xx30 platforms, with SPI space being already saturated, I will give the maintainsership of those platforms, without neutered ME and that space made available for BIOS region to someone else.

I'm out of ideas, like said so many times before under #590 and all tagged issues.

**Comment by tlaurion** [comment]
Al always, I'm open to debate and go forward with constructive solutions.

**Comment by tlaurion** [comment]
x230: @tlaurion @osresearch @merge @jan23 @flawedworld @MrChromebox @shamen123 @eganonoa @bwachter @Thrilleratplay @SebastianMcMillan
t430: @Siproqu

Testers: Ready to merge unless told otherwise. 
You can check building recipe here used for CircleCI: 

- https://github.com/osresearch/heads/pull/703/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R11 and 
- https://github.com/osresearch/heads/pull/703/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R242-R362 

Look at the build results [here](https://app.circleci.com/pipelines/github/tlaurion/heads/623/workflows/118fb09b-f2c1-4348-9ca5-cd68615e7f66/jobs/671)
And download the CircleCI roms produced [here](https://app.circleci.com/pipelines/github/tlaurion/heads/623/workflows/118fb09b-f2c1-4348-9ca5-cd68615e7f66/jobs/671/artifacts)

Planning to merge this at the same time as #912 : 
Ready to merge unless testers say otherwise.
t420: @SebastianMcMillan alexmaloteaux
x220:  @SebastianMcMillan @techge @eganonoa @shamen123 @Thrilleratplay @userongihu @BlackMaria

**Comment by fhvyhjriur** [comment]
It is impossible to unlock the IFD/ME, even with 1vyrain or other things when i only flashed the 4M top-chip with coreboot/skulls? In https://github.com/merge/skulls there is written that flashing the 8M-Chip is optional and that is why i have not flashed it. I also gave my external flasher to other people and dont have it any more because i thought i wont need it any more. Now i am not able to use the heads xx30-maximized images?

**Comment by MrChromebox** [comment]
@fhvyhjriur that's correct. The IFD can be unlocked only via external flashing, and doing so is required to use the -maximized Heads images

**Comment by tlaurion** [comment]
Ok. This is ready to merge!
@MrChromebox ? Some things my eyes missed?
The magic operates [here](https://github.com/osresearch/heads/pull/703/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R245), prior of each maximized board being compiled.
If the user wants to use his own me.bin, gbe.bin and ifd.bin, he can call [extract.sh](https://github.com/osresearch/heads/pull/703/files#diff-a4640f2912da7f1cf5e5f1e805980df6628eed61db884b57513312ec85ee07ca) instead or follow README manual recipe.

File: boards/t430-hotp-maximized/t430-hotp-maximized.config [hunk 2]
```diff
@@ -0,0 +1,86 @@
+# Configuration for a T430 running Qubes and other Linux Based OSes (through kexec)
+#
+# Includes 
+# - Deactivated+neutered ME and expended consequent IFD BIOS regions
```
Comment (flammit) [inline boards/t430-hotp-maximized/t430-hotp-maximized.config:4]: NIT: "expended". I assume you mean expanded or extended.  This typo is in many files, so please get them all.

File: boards/t430-hotp-maximized/t430-hotp-maximized.config [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+export CONFIG_BOOT_REQ_ROLLBACK=n
+export CONFIG_BOOT_KERNEL_ADD="intel_iommu=on intel_iommu=igfx_off"
+export CONFIG_BOOT_KERNEL_REMOVE="quiet"
+export CONFIG_BOOT_DEV="/dev/sda1"
+export CONFIG_BOARD_NAME="Thinkpad T430-hotp-maximized"
+export CONFIG_FLASHROM_OPTIONS="--force --noverify-all -p internal"
+
+# xx30-external-flash boards require of you initially call one of the following to habe gbe.bin ifd.bin and me.bin
+#  - blobs/xx30/download_clean_me.sh
+#     To download Lenovo original ME binary, neuter+deactivate ME, produce reduced IFD ME region and expended BIOS IFD region.
```
Comment (flammit) [inline boards/t430-hotp-maximized/t430-hotp-maximized.config:67]: Same here

File: boards/t430-hotp-maximized/t430-hotp-maximized.config [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+#  - blobs/xx30/extract.sh
+#     To extract from backuped 8M (bottom SPI) ME binary, GBE and 
+# This board has two SPI flash chips, an 8 MB that holds the IFD,
+# the ME image and part of the coreboot image, and a 4 MB one that
+# has the rest of the coreboot and the reset vector.
+#   
+# As a consequence, this replaces the need of having to flash x230-flash and expends available CBFS region (11.5Mb available CBFS space)
```
Comment (flammit) [inline boards/t430-hotp-maximized/t430-hotp-maximized.config:74]: Same here

File: boards/x230-hotp-verification/x230-hotp-verification.config.orig [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# Configuration for a x230-hotp-verification (Nitrokey/Purism USB Security dongle enabled HOTP support)
```
Comment (flammit) [inline boards/x230-hotp-verification/x230-hotp-verification.config.orig:1]: I don't think this file is meant to be included

**Review by flammit — CHANGES_REQUESTED** [review]
Except for the NITs and extra file - this LGTM (without having tested it on real hardware).  

Please also squash the commits before merging there is way too much noise for what is a straight forward blob/boards addition.

**Comment by tlaurion** [comment]
Misses documentation https://github.com/osresearch/heads-wiki/issues/55

**Comment by tlaurion** [comment]
See https://github.com/osresearch/heads-wiki/issues/25#issuecomment-817312900

Make sure to address @qubicrm https://github.com/osresearch/heads-wiki/issues/25#issuecomment-817207672

**Pull request merged**